### PR TITLE
Added Poisson Loss as an objective & Time Distributed Flatten, Conv2D, and MaxPooling2D

### PR DESCRIPTION
**Pull request by anayebi**
I added Poisson loss as an objective. The Poisson loss is the negative log likelihood of data y_true given predictions y_pred, according to a Poisson model (assumes that y_pred > 0). Thus, the loss function is given by y_pred - y_true * log(y_pred).

This loss is typically used in computational neuroscience, when predicting the probability that a neuron will fire. Should be used after a softplus activation, to ensure that y_pred > 0. Often more useful in computational models of cells than the usual cross entropy loss as Poisson loss more heavily penalizes the model for predicting a spike during long periods of cell inactivity, than cross entropy would.

### CODE REVIEW

**Comment by the-moliver** [comment]
You beat me to it! ;-) This will complement https://github.com/fchollet/keras/issues/422 nicely

**Comment by anayebi** [comment]
Great to hear!

I have also added time distributed versions of Flatten (in core.py), Convolution2D (in convolutional.py), and MaxPooling2D (in convolutional.py).

These are useful when connecting a convolutional network with a recurrent net, such as the LSTM, as mentioned in https://github.com/fchollet/keras/issues/401.

Documentation for each layer below:

1. TimeDistributedFlatten
keras.layers.core.TimeDistributedFlatten()

Input shape: (nb_samples, timesteps, *)
Output shape: (nb_samples, timesteps, nb_input units)
Example:
#input shape: (nb_samples, timesteps, 10, 10)
model.add(TimeDistributedFlatten()) #output shape: (nb_samples, timesteps, 100)

2. TimeDistributedConvolution2D
keras.layers.convolutional.TimeDistributedConvolution2D(nb_filter, stack_size, nb_row, nb_col, 
        init='glorot_uniform', activation='linear', weights=None, 
        border_mode='valid', subsample=(1, 1),
        W_regularizer=None, b_regularizer=None, W_constraint=None)

Input shape: 5D Tensor with shape (nb_samples, timesteps, stack_size, nb_row, nb_col)
Output shape: 5D tensor with shape: (nb_samples, timesteps, nb_filter, nb_row, nb_col). nb_row, nb_col might have changed due to padding.

Arguments same as Convolution2D.

3. TimeDistributedMaxPooling2D
keras.layers.convolutional.TimeDistributedMaxPooling2D(poolsize=(2, 2), ignore_border=True)

Input shape: 5D tensor with shape: (nb_samples, timesteps, stack_size, nb_row, nb_col)
Output shape: 5D tensor with shape: (nb_samples, timesteps, stack_size, new_nb_row, new_nb_col)

Arguments same as MaxPooling2D.

**Comment by fchollet** [comment]
> I added Poisson loss as an objective

Ok, we can add that...

> I have also added time distributed versions of Flatten (in core.py), Convolution2D (in convolutional.py), and MaxPooling2D (in convolutional.py).

This is starting to feel impractical. I think a better solution would be a TimeDistributed layer, which would take as argument a base layer, expect an input with one added dim at position 1 (time), and would apply the base layer to the input at every time step.

**Comment by anayebi** [comment]
I agree, and I think having a TimeDistributed layer is a much better design decision (though I'm not exactly sure how you'd prefer to stylistically design it). I wrote the time distributed versions of those three layers because they easily solved my problem of connecting a CNN with any type of RNN, but I imagine a TimeDistributed layer would be implemented in much the same way.

**Comment by usmanatif** [comment]
@anayebi  hello i followed #4172 #421 and tries every possible way to integrate CNN (VGG16) with LSTM in keras, but i am continuously getting error

below is my code

from keras.preprocessing.image import ImageDataGenerator, array_to_img, img_to_array, load_img
from keras.applications.vgg16 import VGG16
from keras.preprocessing import image
from keras.applications.vgg16 import preprocess_input
from keras.models import Sequential
from keras.layers import Flatten, LSTM, Dense, TimeDistributed, InputLayer
from keras.layers.core import Dropout
from keras.layers.convolutional import Convolution2D, MaxPooling2D
import numpy as np

#%%     Image Generator for datasets
datagen = ImageDataGenerator(
        rotation_range = 0,
        width_shift_range = 0,
        height_shift_range = 0,
        shear_range = 0,
        zoom_range = 0,
        rescale = None,
        horizontal_flip = False,
        fill_mode = 'nearest')

train_generator = datagen.flow_from_directory(
        'D:\SpyderDeepLearning\PersonReidentification\Dataset_Light\Train',
        target_size = (224,224),
        batch_size = 1, 
        class_mode = 'categorical')

test_generator = datagen.flow_from_directory(
        'D:\SpyderDeepLearning\PersonReidentification\Dataset_Light\Test',
        target_size = (224,224),
        batch_size = 1, 
        class_mode = 'categorical')


#%%     Convolutional Model (Basically Its VGG) + LSTM model (TimeDistributed)

model = Sequential()

model.add(InputLayer(input_shape=(5, 224, 224, 3)))
model.add(TimeDistributed(Convolution2D(64, (3, 3))))
model.add(TimeDistributed(MaxPooling2D((2,2), strides=(2,2))))
model.add(LSTM(10))
model.add(Dense(3))

model.compile(loss='mean_squared_error', optimizer='rmsprop', metrics=['accuracy'])
model.fit_generator(train_generator, epochs=1, steps_per_epoch=len(train_generator.filenames))
i am getting error that (here 5 is time step, 224,224 is image dimension and 3 is channel )
Error when checking input: expected input_2 to have 5 dimensions, but got array with shape (5, 224, 224, 3)

but when i supply samples also
model.add(InputLayer(input_shape=(1, 5, 224, 224, 3)))
i get this error
number of input channels does not match corresponding dimension of filter, 224 != 3
